epytope/EpitopeSelection/OptiTope.py

```diff
@@ -128,15 +133,26 @@ def __init__(self, results,  threshold=None, k=10, solver="glpk", verbosity=0):
         peps = {}
         cons = {}
 
-        #unstack multiindex df to get normal df based on first prediction method
-        #and filter for binding epitopes
-        method = results.index.values[0][1]
-        res_df = results.xs(results.index.values[0][1], level="Method")
-        res_df = res_df[res_df.apply(lambda x: any(x[a] > self.__thresh.get(a.name, -float("inf"))
+        # unstack multiindex df to get normal df based on first prediction method and score metric
+        method = results.columns[1][1]
+        metric = 1 if rank else 0
+        res_df = copy.deepcopy(results)
+        res_df.columns = res_df.columns.droplevel(1)
+        # Keep only the neceessary information: Allele, and Score/Rank of Peptides
+        # Multiindex structure is broken up here
+        res_df = res_df.xs(res_df.columns.values[metric][1], level="ScoreType", axis=1)
+
+        # and filter for binding epitopes depending on a (high) score or a (low) rank 
+        # with respect to the given threshold
+        if rank:
+            res_df = res_df[res_df.apply(lambda x: any(x[a] < self.__thresh.get(a, float("inf"))
+                                                   for a in res_df.columns), axis=1)]
+        else:
+            res_df = res_df[res_df.apply(lambda x: any(x[a] > self.__thresh.get(a, -float("inf"))
                                                    for a in res_df.columns), axis=1)]
 
         for tup in res_df.itertuples():
-            p = tup[0]
+            p = Peptide(tup[0])
```

Comment:
Ah! 👍 This is because I did not have Peptide Objects stored in the result but Peptide Strings in an older iteration. That can be removed. Checked it locally with `TestOptiTope.py` as well.

Reply:
Perfect!